File: src/jsMain/kotlin/kingdom/lite/ui/KingdomIcon.kt
package kingdom.lite.ui

import kotlinx.browser.document
import kotlinx.browser.window
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.html.a
import kotlinx.html.classes
import kotlinx.html.dom.create
import kotlinx.html.i
import kotlinx.html.js.onClickFunction
import org.w3c.dom.HTMLElement
import org.w3c.dom.asList
import kotlin.js.json

external val Hooks: dynamic

/**
 * Creates the Kingdom icon button for the party sheet
 */
fun createKingdomIcon(actorId: String): HTMLElement {
    return document.create.a {
        classes = setOf("create-button")
        i {
            classes = setOf("fa-solid", "fa-chess-rook")
        }
        attributes["data-tooltip"] = "Open Kingdom Sheet"
        onClickFunction = { event ->
            event.preventDefault()
            event.stopPropagation()
            
            // Open the kingdom UI for this actor
            openKingdomUI(actorId)
        }
    }
}

/**
 * Opens the Kingdom UI using the new KingdomSheet
 */
fun openKingdomUI(actorId: String) {
    console.log("Opening Kingdom UI for actor: $actorId")
    
    // Inject styles if not already done
    injectStyles(KingdomSheetStyles.styles)
    
    // Check if we should use ApplicationV2 (Foundry v10+)
    val hasApplicationV2 = js("typeof foundry !== 'undefined' && foundry.applications && foundry.applications.api && foundry.applications.api.ApplicationV2") as? Boolean ?: false
    
    if (hasApplicationV2) {
        console.log("Using Foundry ApplicationV2")
        openKingdomSheetV2()
    } else {
        console.log("Using custom Application implementation")
        // Fallback to our custom implementation
        kotlinx.coroutines.GlobalScope.launch {
            val sheet = KingdomSheet()
            sheet.render(true)
        }
    }
}

/**
 * Registers the hook to add Kingdom icons to party actors in the sidebar
 */
fun registerKingdomIconHook() {
    console.log("Registering Kingdom icon hook")
    
    // Hook into the actor directory render to add our icons
    Hooks.on("renderActorDirectory") { app: Any, html: dynamic, data: Any ->
        console.log("renderActorDirectory hook triggered")
        
        // Cast html to HTMLElement
        val htmlElement = html as HTMLElement
        
        // Find all party actor folders in the sidebar
        htmlElement.querySelectorAll(".folder[data-party]")
            .asList()
            .filterIsInstance<HTMLElement>()
            .forEach { folderElement ->
                // Access the dataset directly through dynamic cast
                val datasetElement = folderElement.asDynamic()
                val actorId = datasetElement.dataset.entryId as String?
                
                if (actorId != null) {
                    console.log("Found party actor folder with ID: $actorId")
                    
                    // Find where to insert our icon (after the folder name)
                    val folderName = folderElement.querySelector(".folder-name") as? HTMLElement
                    if (folderName != null) {
                        // Insert the Kingdom icon
                        val kingdomIcon = createKingdomIcon(actorId)
                        folderName.insertAdjacentElement("afterend", kingdomIcon)
                        console.log("Inserted Kingdom icon for actor: $actorId")
                    }
                }
            }
        
        Unit
    }
}

File: src/ui/KingdomIcon.ts
// Kingdom Icon module for PF2e ReignMaker
// Adds Kingdom buttons to party actors in the sidebar

// Import PF2e types
/// <reference path="../types/pf2e-types.d.ts" />

// Import KingdomApp directly instead of dynamic import
import { KingdomApp } from '../view/kingdom/KingdomApp';
import { logger } from '../utils/Logger';

/**
 * Creates the Kingdom icon button for the party sheet
 */
export function createKingdomIcon(actorId: string): HTMLElement {
    const link = document.createElement('a');
    link.classList.add('create-button');
    link.setAttribute('data-tooltip', 'ReignMaker');
    
    const icon = document.createElement('i');
    icon.classList.add('fa-solid', 'fa-chess-rook');
    link.appendChild(icon);
    
    link.addEventListener('click', (event) => {
        event.preventDefault();
        event.stopPropagation();
        
        // Open the kingdom UI for this actor
        openKingdomUI(actorId);
    });
    
    return link;
}

/**
 * Opens the Kingdom UI using the new Svelte KingdomApp
 */
export function openKingdomUI(actorId: string): void {


    try {
        const app = new KingdomApp({ actorId } as any);

        app.render(true, { focus: true });

    } catch (error) {
        logger.error("PF2e ReignMaker | Failed to create KingdomApp:", error);
        // @ts-ignore
        ui.notifications?.error("Failed to open Kingdom management UI");
    }
}

/**
 * Registers the hook to add Kingdom icons to party actors in the sidebar
 */
export function registerKingdomIconHook(): void {

    // Hook into the actor directory render to add our icons
    Hooks.on("renderActorDirectory", (app: any, html: any, data: any) => {

        // Convert jQuery to HTMLElement if needed
        const htmlElement = html instanceof HTMLElement ? html : html.get(0);
        
        if (!htmlElement) {

            return;
        }
        
        // Try multiple selectors for party actors (for compatibility with different PF2E versions)
        const partySelectors = [
            ".folder[data-party]",
            "[data-folder-id] .directory-item[data-actor-type='party']",
            ".directory-item[data-document-id]"
        ];
        
        let foundPartyActors = false;
        
        for (const selector of partySelectors) {
            const elements = htmlElement.querySelectorAll(selector);
            
            if (elements.length > 0) {
                elements.forEach((element: Element) => {
                    const elem = element as HTMLElement;
                    const actorId = elem.dataset.entryId || elem.dataset.documentId || elem.dataset.actorId;
                    
                    if (actorId) {
                        // @ts-ignore
                        const actor = game.actors?.get(actorId);
                        
                        if (actor && actor.type === 'party') {
                            foundPartyActors = true;

                            // Find where to insert our icon
                            const nameElement = elem.querySelector(".document-name, .entity-name, .folder-name");
                            
                            if (nameElement && !elem.querySelector('.fa-chess-rook')) {
                                // Insert the Kingdom icon
                                const kingdomIcon = createKingdomIcon(actorId);
                                nameElement.insertAdjacentElement("afterend", kingdomIcon);

                            }
                        }
                    }
                });
            }
        }
        
        if (!foundPartyActors) {

            // Fallback: check all actors for party type
            // @ts-ignore
            const partyActors = game.actors?.filter((a: any) => a.type === 'party');
            if (partyActors && partyActors.length > 0) {

                partyActors.forEach((actor: any) => {
                    // Try to find the element for this actor
                    const actorElem = htmlElement.querySelector(
                        `[data-entry-id="${actor.id}"], [data-document-id="${actor.id}"], [data-actor-id="${actor.id}"]`
                    );
                    
                    if (actorElem) {
                        const nameElement = actorElem.querySelector(".document-name, .entity-name");
                        
                        if (nameElement && !actorElem.querySelector('.fa-chess-rook')) {
                            const kingdomIcon = createKingdomIcon(actor.id);
                            nameElement.insertAdjacentElement("afterend", kingdomIcon);

                        }
                    }
                });
            }
        }
    });
}

/**
 * Initialize the Kingdom Icon module
 */
export function initKingdomIcon(): void {
    // Register hooks immediately - this is called from index.ts before the ready hook

    if (typeof Hooks !== 'undefined') {
        // Register the hook to add icons when the actor directory is rendered
        registerKingdomIconHook();
    }
}
